File: search-service/internal/registry/object_class_hierarchy_registry.go
package registry

import (
	"github.com/quantara/search-service/internal/domain"
)

type ObjectClassHierarchyRegistry struct {
	parentsOrSelf        map[*domain.ObjectClassInfo]map[*domain.ObjectClassInfo]struct{}
	rootUnderBase        map[*domain.ObjectClassInfo]*domain.ObjectClassInfo
	actualClassesForRoot map[*domain.ObjectClassInfo][]*domain.ObjectClassInfo
}

func NewObjectClassHierarchyRegistry(cfg *domain.SearchDomainConfig, classRegistry *ObjectClassRegistry) *ObjectClassHierarchyRegistry {
	registry := &ObjectClassHierarchyRegistry{
		parentsOrSelf:        make(map[*domain.ObjectClassInfo]map[*domain.ObjectClassInfo]struct{}),
		rootUnderBase:        make(map[*domain.ObjectClassInfo]*domain.ObjectClassInfo),
		actualClassesForRoot: make(map[*domain.ObjectClassInfo][]*domain.ObjectClassInfo),
	}

	if cfg == nil || cfg.Hierarchy == nil || classRegistry == nil {
		return registry
	}

	for childName, parentNames := range cfg.Hierarchy.ParentsOrSelf {
		child := classRegistry.ByName(childName)
		if child == nil {
			continue
		}
		set := make(map[*domain.ObjectClassInfo]struct{})
		for _, parentName := range parentNames {
			parent := classRegistry.ByName(parentName)
			if parent != nil {
				set[parent] = struct{}{}
			}
		}
		registry.parentsOrSelf[child] = set
	}

	for childName, rootName := range cfg.Hierarchy.RootUnderBase {
		child := classRegistry.ByName(childName)
		root := classRegistry.ByName(rootName)
		if child != nil && root != nil {
			registry.rootUnderBase[child] = root
		}
	}

	for rootName, actualNames := range cfg.Hierarchy.ActualClassesForRoot {
		root := classRegistry.ByName(rootName)
		if root == nil {
			continue
		}
		actuals := make([]*domain.ObjectClassInfo, 0, len(actualNames))
		for _, name := range actualNames {
			actual := classRegistry.ByName(name)
			if actual != nil {
				actuals = append(actuals, actual)
			}
		}
		registry.actualClassesForRoot[root] = actuals
	}

	return registry
}

func (r *ObjectClassHierarchyRegistry) IsParentOrSelf(parent, child *domain.ObjectClassInfo) bool {
	if r == nil || parent == nil || child == nil {
		return false
	}
	set := r.parentsOrSelf[child]
	if set == nil {
		return false
	}
	_, ok := set[parent]
	return ok
}

func (r *ObjectClassHierarchyRegistry) ParentsOrSelf(objectClass *domain.ObjectClassInfo) []*domain.ObjectClassInfo {
	if r == nil || objectClass == nil {
		return nil
	}
	set := r.parentsOrSelf[objectClass]
	if set == nil {
		return nil
	}
	result := make([]*domain.ObjectClassInfo, 0, len(set))
	for class := range set {
		result = append(result, class)
	}
	return result
}

func (r *ObjectClassHierarchyRegistry) RootUnderBase(objectClass *domain.ObjectClassInfo) *domain.ObjectClassInfo {
	if r == nil || objectClass == nil {
		return objectClass
	}
	if root, ok := r.rootUnderBase[objectClass]; ok {
		return root
	}
	return objectClass
}

func (r *ObjectClassHierarchyRegistry) ActualClassesForRoot(rootClass *domain.ObjectClassInfo) []*domain.ObjectClassInfo {
	if r == nil || rootClass == nil {
		return nil
	}
	actual := r.actualClassesForRoot[rootClass]
	result := make([]*domain.ObjectClassInfo, 0, len(actual))
	result = append(result, actual...)
	return result
}

File: enricher-service/src/main/java/com/enricher/service/registry/ObjectClassHierarchyRegistry.java
package com.enricher.service.registry;

import com.enricher.service.domain.MetadataExportV3;
import com.enricher.service.domain.ObjectClassInfo;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

@Component
public class ObjectClassHierarchyRegistry {
    private final Map<ObjectClassInfo, Set<ObjectClassInfo>> parentsOrSelf;
    private final Map<ObjectClassInfo, List<ObjectClassInfo>> parentsOrSelfOrdered;
    private final Map<ObjectClassInfo, Set<ObjectClassInfo>> descendantsOrSelfByClass;

    public ObjectClassHierarchyRegistry(MetadataExportV3 metadata, ObjectClassRegistry objectClassRegistry) {
        this.parentsOrSelf = new HashMap<>();
        this.parentsOrSelfOrdered = new HashMap<>();
        this.descendantsOrSelfByClass = new HashMap<>();

        if (metadata == null || metadata.hierarchy() == null || metadata.hierarchy().parentsOrSelf() == null) {
            return;
        }

        Map<ObjectClassInfo, Set<ObjectClassInfo>> descendantsOrSelf = new HashMap<>();
        for (ObjectClassInfo objectClass : objectClassRegistry.all()) {
            descendantsOrSelf.put(objectClass, new HashSet<>(Set.of(objectClass)));
        }

        for (Map.Entry<String, List<String>> entry : metadata.hierarchy().parentsOrSelf().entrySet()) {
            ObjectClassInfo child = objectClassRegistry.byName(entry.getKey());
            if (child == null) {
                continue;
            }
            Set<ObjectClassInfo> parentSet = new HashSet<>();
            List<ObjectClassInfo> ordered = new ArrayList<>();
            if (entry.getValue() != null) {
                for (String className : entry.getValue()) {
                    ObjectClassInfo parent = objectClassRegistry.byName(className);
                    if (parent == null) {
                        continue;
                    }
                    parentSet.add(parent);
                    ordered.add(parent);
                }
            }
            parentsOrSelf.put(child, Set.copyOf(parentSet));
            parentsOrSelfOrdered.put(child, List.copyOf(ordered));

            for (ObjectClassInfo parent : parentSet) {
                if (parent != child) {
                    descendantsOrSelf.get(parent).add(child);
                }
            }
        }

        for (ObjectClassInfo objectClass : objectClassRegistry.all()) {
            descendantsOrSelfByClass.put(objectClass, Set.copyOf(descendantsOrSelf.get(objectClass)));
        }
    }

    public boolean isParentOrSelf(ObjectClassInfo parent, ObjectClassInfo child) {
        if (parent == null || child == null) {
            return false;
        }
        Set<ObjectClassInfo> set = parentsOrSelf.get(child);
        return set != null && set.contains(parent);
    }

    public List<ObjectClassInfo> parentsOrSelfOrdered(ObjectClassInfo objectClass) {
        if (objectClass == null) {
            return List.of();
        }
        List<ObjectClassInfo> chain = parentsOrSelfOrdered.get(objectClass);
        if (chain == null || chain.isEmpty()) {
            return List.of(objectClass);
        }
        return List.copyOf(chain);
    }

    public Set<ObjectClassInfo> descendantsOrSelf(ObjectClassInfo objectClass) {
        if (objectClass == null) {
            return Set.of();
        }
        Set<ObjectClassInfo> result = descendantsOrSelfByClass.get(objectClass);
        if (result == null) {
            throw new IllegalArgumentException("Unknown object class: [" + objectClass.name() + "]");
        }
        return result;
    }
}
